File: modules/data/tenvx.lua
sptbl["tenvx"] = {

    files = {
        module = "tenvx.c",
        header = "tenvx.h",
        example = "ex_tenvx.c",
    },

    func = {
        create = "ut_tenvx_create",
        destroy = "ut_tenvx_destroy",
        init = "ut_tenvx_init",
        compute = "ut_tenvx_compute",
    },

    params = {
        optional = {
            {
                name = "atk",
                type = "UTFLOAT",
                description = "Tau attack time, in seconds. Must be non-zero.",
                default = 0.1
            },
            {
                name = "hold",
                type = "UTFLOAT",
                description = "Hold time, in seconds. The duration of the gate signal.",
                default = 0.3
            },
            {
                name = "rel",
                type = "UTFLOAT",
                description = "Tau release time, in seconds. Must be non-zero.",
                default = 0.2
            },
        }
    },

    modtype = "module",

    description = [[Trigger based exponential AHD envelope generator.
    This envelope generator emulates the exponential behavior of analogue 
envelope generators by passing a gate signal (whose duration is specified via
the hold parameter) through a one-pole filter, whose filter coefficeints are
calculated in terms of tau.  
    ]],

    ninputs = 1,
    noutputs = 1,

    inputs = {
        {
            name = "trig",
            description = "Trigger input. When non-zero, the envelope will (re)trigger."
        },
    },

    outputs = {
        {
            name = "out",
            description = "Signal output"
        },
    }

}

File: modules/tenvx.c
#include <stdlib.h>
#include <math.h>
#include "utone.h"

int ut_tenvx_create(ut_tenvx **p)
{
    *p = malloc(sizeof(ut_tenvx));
    return UT_OK;
}

int ut_tenvx_destroy(ut_tenvx **p)
{
    free(*p);
    return UT_OK;
}


int ut_tenvx_init(ut_data *ut, ut_tenvx *p)
{
    p->hold = 0.5;
    p->atk = 0.5;
    p->rel = 0.5;
    p->a_a = p->b_a = 0;
    p->a_r = p->b_r = 0;
    p->y = 0;
    p->count = (uint32_t) (p->hold * ut->sr);
    return UT_OK;
}


int ut_tenvx_compute(ut_data *ut, ut_tenvx *p, UTFLOAT *in, UTFLOAT *out)
{
    *out = 0;

    if(*in != 0) {
        p->a_a = exp(-1.0/(p->atk * ut->sr));
        p->b_a = 1.0 - p->a_a;
        p->a_r = exp(-1.0/(p->rel * ut->sr));
        p->b_r = 1.0 - p->a_r;
        p->count = (uint32_t) (p->hold * ut->sr);
    }

    if(p->count > 0) {
        *out = p->b_a + p->a_a * p->y;
        p->y = *out;
        p->count--;
    } else {
        *out = p->a_r * p->y;
        p->y = *out;
    }

    return UT_OK;
}
